Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,182 @@
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+import SwiftParserDiagnostics
+
+actor DiagnosticReportManager {
+  /// A task that gets report from sourcekitd
+  private typealias ReportTask = Task<RelatedFullDocumentDiagnosticReport, Error>
+
+  private let sourcekitd: SourceKitD
+  private let syntaxTreeManager: SyntaxTreeManager
+  private let documentManager: DocumentManager
+  private let clientHasDiagnosticsCodeDescriptionSupport: Bool
+
+  private nonisolated var keys: sourcekitd_keys { return sourcekitd.keys }
+  private nonisolated var requests: sourcekitd_requests { return sourcekitd.requests }
+
+  /// The cache that stores reportTasks for snapshot id and buildSettings
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportTaskCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
+      buildSettings: SwiftCompileCommand?,
+      reportTask: ReportTask
+    )] = []
+
+  /// The number of reportTasks to keep
+  ///
+  /// - Note: This has been chosen without scientific measurements.
+  private let cacheSize = 5
+
+  init(
+    sourcekitd: SourceKitD,
+    syntaxTreeManager: SyntaxTreeManager,
+    documentManager: DocumentManager,
+    clientHasDiagnosticsCodeDescriptionSupport: Bool
+  ) {
+    self.sourcekitd = sourcekitd
+    self.syntaxTreeManager = syntaxTreeManager
+    self.documentManager = documentManager
+    self.clientHasDiagnosticsCodeDescriptionSupport = clientHasDiagnosticsCodeDescriptionSupport
+  }
+
+  func diagnosticReport(
+    for snapshot: DocumentSnapshot,
+    buildSettings: SwiftCompileCommand?
+  ) async throws -> RelatedFullDocumentDiagnosticReport {
+    if let report = try await reportTask(for: snapshot.id, buildSettings: buildSettings)?.value {
+      return report
+    }
+    let reportTask: Task<RelatedFullDocumentDiagnosticReport, Error>
+    if let buildSettings, !buildSettings.isFallback {
+      reportTask = Task {
+        return try await requestReport(with: snapshot, compilerArgs: buildSettings.compilerArgs)
+      }
+    } else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we \(buildSettings != nil ? "have fallback build settings" : "don't have build settings")"
+      )
+      // If we don't have build settings or we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      reportTask = Task {
+        return try await requestFallbackReport(with: snapshot)
+      }
+    }
+    setReportTask(for: snapshot.id, buildSettings: buildSettings, reportTask: reportTask)
+    return try await reportTask.value
+  }
+
+  func resetCache() async {
```

Comment:
Why do we ever need to reset the cache? The snapshot ID + build settings should be sufficient that we re-request diagnostic whenever they might have changed. Or am I missing something?

Reply:
there are two tests will be failed without reseting cache.
```
Test Case '-[SourceKitLSPTests.SKTests testDependenciesUpdatedSwift]' failed (5.170 seconds).
Test Case '-[SourceKitLSPTests.LocalSwiftTests testDiagnosticsReopen]' failed (1.400 seconds).
```

i think that's because the cache for the file becomes dirty if the file is reopened with changed dependencies or content.

Reply:
Ah, I see. That makes sense. But we should not need to reset the entire cache, we should only need to clear the diagnostics for the document that we are opening/closing/reopening.

Reply:
No need to apologize., thank you for fixing this issue!